### ISSUE TITLE
Shell.setLocation doesn't work properly (or at all)

### ISSUE THREAD

**Issue by merks** [eclipse-platform/eclipse.platform.swt#790]
**Describe the bug**

Calling Shell.setLocation doesn't visibly move the shell to the specified location, and sometimes does move the shell at all.

**To Reproduce**

I modified this snippet:
```
/*******************************************************************************
 * Copyright (c) 2000, 2005 IBM Corporation and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     IBM Corporation - initial API and implementation
 *******************************************************************************/
package org.eclipse.swt.snippets;

/*
 * Monitor example snippet: center a shell on the primary monitor
 *
 * For a list of all SWT example snippets see
 * http://www.eclipse.org/swt/snippets/
 *
 * @since 3.0
 */
import org.eclipse.swt.graphics.*;
import org.eclipse.swt.widgets.*;

public class Snippet120 {

public static void main (String [] args) {
	Display display = new Display ();
	Shell shell = new Shell (display);
	shell.setText("Snippet 120");
	shell.setSize (200, 200);
	Monitor primary = display.getPrimaryMonitor ();
	Rectangle bounds = primary.getBounds ();
	Rectangle rect = shell.getBounds ();
	System.out.println("primaryMnitorBounds=" + bounds);
	int x = bounds.x + (bounds.width - rect.width) / 2;
	int y = bounds.y + (bounds.height - rect.height) / 2;
	String gtkVersion = System.getProperty("org.eclipse.swt.internal.gtk.version");
	System.out.println("gtkVersion=" + gtkVersion);
	System.out.println("location=" + x + "," + y);
	shell.setLocation (x, y);
	System.out.println("location=" + shell.getLocation());
	shell.open ();
	while (!shell.isDisposed ()) {
		if (!display.readAndDispatch ()) display.sleep ();
	}
	display.dispose ();
}
}
```
Here we see it run and clearly the location is not in the center of the monitor though it says it's in the center:

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/208716/8b22abb4-a62e-4797-a9c7-b91055b31a7a)

The problem with the Eclipse Installer is even worse, where I cannot move the installer as we see here the value after calling set location remains unchanged:

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/208716/df2dfb9e-1189-418f-9469-3c40bfea6a43)

And I cannot open the "menu" in the correct location:

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/208716/9bd2279c-5492-464f-879b-35f6fb17d5b9)

**Environment:**

1. Select the platform(s) on which the behavior is seen:
- - [ ] All OS
- - [ ] Windows
- - [x] Linux  Fedora 38
- - [ ] macOS 

2. Additional OS info (e.g. OS version, Linux Desktop, etc)

I'm happy to provide details.  I installed this to test Wayland, but I don't know if that's what I have.  I have tried to install GTK4 but SWT still uses gtkVersion=3.24.37.

4. JRE/JDK version

jdk-17.0.8+7

5.  What to do?

I'd be happy to find out that I can configure my OS so that SWT uses GTK4 and that solves the problem or that I can do something else in this OS to work around the problem.

Note that even in the IDE, Ctrl-T opens the dialog in a strange location:

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/208716/42e33abd-494e-42c5-add4-9d4291af3e13)

**Comment by deepika-u** [comment]
Hi, I tried around 3 instances on my windows 11 and dont see any such problem.

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/46807064/170dec95-e145-47d2-87cc-2e9233544cc4)

When shell size is 400,400
![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/46807064/ffe22ebd-5b46-4124-a331-e47bc883c8df)

For ctrl+T
![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/46807064/a4eaef8b-9cd9-4d85-a3c1-3409f3373bbf)

Add 200 for both the x and y bounds, this also comes up properly as shown below
![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/46807064/bcb6c44c-06ed-4e4d-80ad-9eeae5ef3a35)


I tried on the below environment 
Eclipse SDK
Version: 2023-09 (4.29)
Build id: I20230829-1800
OS: Windows 11, v.10.0, x86_64 / win32
Java vendor: Eclipse Adoptium
Java runtime version: 20+36
Java version: 20

**Comment by merks** [comment]
Of course it works on Windows.  The Eclipse Installer also works very well on Windows.  That's why I marked it as only a Linux problem.  Probably a Wayland problem.  I'm going to change the title to be more clear on this being a Linux problem.

**Comment by Phillipus** [comment]
>  Probably a Wayland problem

Yes.

**Comment by akurtakov** [comment]
Wayland protocol does not allow clients to position top level surfaces, and there is no global screen coordinates system. The move() API in GTK3 does not work under Wayland, and it has been removed in GTK4. Note the missing "gtk_widget_move" in GTK4 https://developer-old.gnome.org/gtk4/stable/GtkWidget.html .

**Comment by mickaelistria** [comment]
Just an idea here: could thing work better if Shell have a parent? Wayland usually allows parent-relative positioning, so if a parent is set in GTK, it might be possible to position a child widget. That would mean that on Wayland, position works only if Shell has a parent and coordinates would have to be relative (so we may need either conversion or new API)

**Comment by merks** [comment]
> Just an idea here: could thing work better if Shell have a parent? Wayland usually allows parent-relative positioning, so if a parent is set in GTK, it might be possible to position a child widget. That would mean that on Wayland, position works only if Shell has a parent and coordinates would have to be relative (so we may need either conversion or new API)

The installer's menu shell is already a child of the installer dialog shell:

![image](https://github.com/user-attachments/assets/7f086531-9e7c-4f6e-9e39-27dad62284cf)

**Comment by HeikoKlare** [comment]
Setting a parent shell is not sufficient (at least with the current SWT implementation). The misplaced shells I have seen so far (popups, dialogs like "open type", find/replace overlay) all have a parent shell and are still not placed properly. If I remember correctly, the only solution I found is to initialize the shell with `SWT.ON_TOP`. But maybe there is also something that can be improved in the SWT `Shell` implementation that allows for positioning of other kinds of child shells?

**Comment by mickaelistria** [comment]
>  If I remember correctly, the only solution I found is to initialize the shell with SWT.ON_TOP

If this works, that's promising. Do you know how having "ON_TOP" leads SWT to perform different GTK calls that work with wayland?

**Comment by merks** [comment]
Yes, I tried this flag with the installer, and indeed the menu shell opens in the right place, which is an improvement, but then, as you alluded, I can't push the menu button again to make the menu dismiss, I have to hit the x button to close the menu, so it's kind of a trade-off with the random location of the menu.  In any case, the installer dialog can't be moved because it's the root.

 All the dialogs in the IDE that open in random places is just beyond annoying...

**Comment by ep9io** [comment]
Try calling `export GDK_BACKEND=x11` before starting up the process.  I found that help sometimes.  I'm on Fedora 39 Wayland.

**Comment by Phillipus** [comment]
> Try calling `export GDK_BACKEND=x11` before starting up the process. I found that help sometimes. I'm on Fedora 39 Wayland.

If you set that then Eclipse is not running on Wayland. The point of this issue is about setting the location of a Shell on Wayland.

**Comment by ep9io** [comment]
> > Try calling `export GDK_BACKEND=x11` before starting up the process. I found that help sometimes. I'm on Fedora 39 Wayland.
> 
> If you set that then Eclipse is not running on Wayland. The point of this issue is about setting the location of a Shell on Wayland.

My apologies. I was only providing a workaround; as you stated, the issue regarding Wayland still exists.

**Comment by akurtakov** [comment]
As this is Wayland design decision there is nothing SWT can do.
